Scroll towards certain grid edge if user hovers on sticky

Michota opened the pull request:
Its **possible fix may be required**
It may be a bug, or a feature, but **to scroll towards certain direction user has to constantly move the pointer.**

Michota committed:
git pull v5 dev

Michota committed:
Merge branch 'v5-dev' of https://github.com/silevis/reactgrid into v5-dev-advanced-selection

Michota committed:
Merge branch 'v5-dev' of https://github.com/silevis/reactgrid into v5-dev-cell-measure

Michota committed:
feat: Add getCellFromPointer utility function

Michota committed:
feat: Add getCellPane, getStickyPaneDirection, isCellSticky

Michota committed:
feat: add calcScrollBy utility function

Michota committed:
chore: comment-out and format code

Michota committed:
feat: replace getLocationFromClient with getCellFromPointer

Michota committed:
feat: create getRowsAndColumns function

Michota committed:
feat: Add CellsMeasurements interface to PublicModel.ts

Michota committed:
fix: make corner-stickies selection act as other stickies

- calculate selecting speed based on distance
- refactor sticky cell case block
- create helper functions

Michota committed:
test: crate scrollTowardsSticky function

- bug: it wont move more than once if new cell under sticky is selected

Michota committed:
refactor: extract scrolling feature out of non-related to it function

Michota committed:
Merge branch 'v5-dev-scrolltowardsticky' of https://github.com/silevis/reactgrid into v5-dev-scrolltowardsticky

Michota committed:
minor: remove unnescesarry import

Michota committed:
minor: remove unnescesarry import

Michota committed:
Merge branch 'v5-dev-scrolltowardsticky' of https://github.com/silevis/reactgrid into v5-dev-scrolltowardsticky

Michota committed:
Merge branch 'v5-dev-scrolltowardsticky' of https://github.com/silevis/reactgrid into v5-dev-scrolltowardsticky

Michota committed:
Merge branch 'v5-dev-scrolltowardsticky' of https://github.com/silevis/reactgrid into v5-dev-scrolltowardsticky

Michota committed:
Merge branch 'v5-dev-scrolltowardsticky' of https://github.com/silevis/reactgrid into v5-dev-scrolltowardsticky

Michota committed:
Merge branch 'v5-dev-scrolltowardsticky' of https://github.com/silevis/reactgrid into v5-dev-scrolltowardsticky

Michota committed:
docs: add jsdocs

Michota committed:
fix: scroll towards end of grid-border if pointer hovers sticky

Michota committed:
test: temporary add interval to handle select and scroll

Michota committed:
chore: remove unnecessary code, add Jsdocs

DLowHP commented on src/types/PublicModel.ts:
```diff
@@ -94,3 +94,58 @@ export interface ReactGridProps {
+     */
+    offsetsLeft: number[];
+  }[];
+}
```
This belongs to internal model; I would move this to a seperate file `CellsMeasurements.ts`, especially given the fact that we are not using this at the moment.

DLowHP commented on src/utils/calcScrollBy.ts:
```diff
@@ -0,0 +1,52 @@
+export const calcScrollBy = (direction: string, multiplier?: number): { x: number; y: number; } => {
+  let x = 0;
+  let y = 0;
+  switch (direction.toLowerCase()) {
```
Do we need `toLowerCase()`?

Michota replied:
It won't do harm, but may make someones life easier.

DLowHP commented on src/utils/calcScrollBy.ts:
```diff
@@ -0,0 +1,52 @@
+ * @param multiplier The multiplier to apply to the scroll direction.
+ * @returns The scroll direction.
+ */
+export const calcScrollBy = (direction: string, multiplier?: number): { x: number; y: number; } => {
```
You can type `direction` as `PaneName` from `InternalModel.ts`

```suggestion
import { PaneName } from "../types/InternalModel";

/**
 * Calculates the scroll direction. It also can apply a multiplier to the scroll speed in selected direction.
 * @param direction The direction to scroll.
 * @param multiplier The multiplier to apply to the scroll direction.
 * @returns The scroll direction.
 */
export const calcScrollBy = (direction: PaneName, multiplier?: number): { x: number; y: number; } => {
```

DLowHP commented on src/utils/cellUtils.ts:
```diff
@@ -1,65 +1,70 @@
+import { Columns } from "../blueprint";
```
Did you mean `Column` from `PublicModel`? Or was this import just unintentional?

DLowHP commented on src/utils/cellUtils.ts:
```diff
@@ -96,4 +101,37 @@ export const findMinimalSelectedArea = (store: ReactGridStore, currentArea: Nume
+ * @returns True if the cell is on sticky pane, false otherwise.
+ */
+export function isCellSticky(store: ReactGridStore, cell: Cell): boolean {
+  return getStickyPaneDirection(getCellPane(store, cell))?.toLowerCase() === "center" ? false : true;
```
I believe this could be optimized and potentially simplifed if we'd just use the `isCellInRange` function.

```suggestion
  return !isCellInRange(store, cell, store.paneRanges.Center);
```

Why I think `isCellInRange` could be more performant?: the functions used currently are calling the DOM, also traversing the parent elements while the `isCellInRange` only works on the store objects.

DLowHP commented on src/utils/cellUtils.ts:
```diff
@@ -96,4 +101,37 @@ export const findMinimalSelectedArea = (store: ReactGridStore, currentArea: Nume
+export function getStickyPaneDirection(pane: HTMLElement): string | undefined {
+  const direction = [...pane.classList].find((className) => className.includes("rgPane-"))?.replace("rgPane-", "");
+  return direction;
+}
```
This seems to be a little overkill; I think this two functions can be replaced with `isCellInRange` by "matching" the provided cell to the `isCellInRange` call that returns `true`.

DLowHP commented on src/utils/createMultiplierFromDistance.ts:
```diff
@@ -0,0 +1,5 @@
+
```
It won't hurt to inform future visitors that this uses the "Euclidean distance" formula ;)

```suggestion
/**
 * This function calculates the Euclidean distance between two points in a 2D space.
 * The distance is calculated using the formula sqrt((aX + aY)^2 + (bX + bY)^2).
 *
 * @param aX - The x-coordinate of the first point.
 * @param aY - The y-coordinate of the first point.
 * @param bX - The x-coordinate of the second point.
 * @param bY - The y-coordinate of the second point.
 * @returns The Euclidean distance between the two points.
 */
```

DLowHP commented on src/utils/getCellFromCertainPane.ts:
```diff
@@ -0,0 +1,23 @@
+    const pane = getCellPane(store, store.getCellByIndexes(rowIndex, colIndex)!);
+
+    if (getStickyPaneDirection(pane)?.toLowerCase() === paneName.toLowerCase()) cell = container;
+  });
```
I think it will be better to use `find()` returning the cell as soon as it's found and I'd rename the `rowsAndCols` to `cellIndexes`:

```suggestion
  const cell = cellContainers.find((container) => {
    const cellIndexes = getRowAndColumns(container);
    if (!cellIndexes) return;

    const { rowIndex, colIndex } = cellIndexes;

    const pane = getCellPane(store, store.getCellByIndexes(rowIndex, colIndex)!);

    return getStickyPaneDirection(pane)?.toLowerCase() === paneName.toLowerCase();
  });
```

DLowHP commented on src/utils/getCellFromPointer.ts:
```diff
@@ -0,0 +1,32 @@
+ * @returns The row and column index of the cell that contains the pointer
+ */
+
+export const getCellFromPointer = (
```
If this only returns the indexes and not the Cell object it'd be better named `getCellFromPointerIndexes`:
```suggestion
export const getCellFromPointerIndexes = (
```

DLowHP replied:
Geez, sorry for the wrong suggestion but this does not make sense in English as I got the order wrong - it should be named getCellIndexesFromPointer(Location)... I'll fix this myself in a moment, my bad! 🤦🏻‍♂️

DLowHP commented on src/utils/getCellFromPointer.ts:
```diff
@@ -0,0 +1,32 @@
+
+  if (!rowsAndColumns) return defaultReturn;
+
+  const { rowIndex, colIndex } = rowsAndColumns;
```
```suggestion
  const cellIndexes = getRowAndColumns(cellContainer);

  if (!cellIndexes) return defaultReturn;

  const { rowIndex, colIndex } = cellIndexes;
```

DLowHP commented on src/utils/getCellFromPointer.ts:
```diff
@@ -0,0 +1,32 @@
+  clientY: number
+): {  rowIndex: -1; colIndex: -1 } | { rowIndex: number; colIndex: number } => {
+  // Return if no expectations were met
+  const defaultReturn = {  rowIndex: -1, colIndex: -1 };
```
I'm not sure what's the best name for this but I'd suggest `emptyIndexes` or `noCellIndexes` instead of `defaultReturn`.

DLowHP commented on src/utils/getRowAndColumns.ts:
```diff
@@ -0,0 +1,18 @@
+ * @returns The row and column index of the cell container element.
+ */
+
+export function getRowAndColumns(cellContainer: Element): { rowIndex: number; colIndex: number; } | false {
```
Per previous suggestions, I'd rename this to `getCellIndexes` because `getRowAndColumns` suggests that this returns `Row` and `Column` objects or something similar which might be confusing. 
```suggestion
export function getCellIndexes(cellContainer: Element): { rowIndex: number; colIndex: number; } | false {
```

DLowHP commented on src/utils/createMultiplierFromDistance.ts:
```diff
@@ -0,0 +1,5 @@
+
+export function createMultiplierFromDistance(aX: number, aY: number, bX: number, bY: number) {
```
Also, I think that this should be named `getDistanceBetweenTwoPoints` because that's essentially what this does - returns the distance not the multiplier. It _is_ used to get a multiplier as I saw later, but this function is more generic and I believe the name should also be more generic.

Michota committed:
Update src/utils/getCellFromCertainPane.ts

Co-authored-by: Daniel Fidor <daniel.fidor@silevis.com>

Michota committed:
Update src/utils/getRowAndColumns.ts

Co-authored-by: Daniel Fidor <daniel.fidor@silevis.com>

Michota committed:
style: make code more accessible

Michota committed:
Merge branch 'v5-dev-scrolltowardsticky' of https://github.com/silevis/reactgrid into v5-dev-scrolltowardsticky

DLowHP committed:
Merge branch 'v5-dev-scrolltowardsticky' of https://github.com/silevis/reactgrid into v5-dev-scrolltowardsticky

DLowHP committed:
renamed getCellFromPointerIndexes to getCellIndexesFromPointerLocation

DLowHP committed:
updated getCellIndexes
Renamed aforementioned function to getCellIndexesFromContainerElement, updated uses to match new name, removed mistakenly merged old interval code

DLowHP committed:
simplified getCellFromCertainPane
using isCellInRange instead getCellPane + getStickyPaneDirection to match cell's parent pane, set stricter type for paneName, updated getNonStickyCell paneName argument to match type,  changed return type to match usage
